### DISCUSSION TITLE
How to get the deployed contract address.

### DISCUSSION THREAD

**Discussion by Abhishek-1857** [hyperledger-labs/solang#750] [General]
hello i deployed a contract written in solidity on testnet . how can i access the contract address.

**Comment by LucasSte** [comment]
Hey Abhishek, which blockchain are you targeting?

**Reply by Abhishek-1857** [reply]
solana only

**Comment by seanyoung** [comment]
@Abhishek-1857 could you please post steps of what are doing and what exactly you would like to do? There are a few different things you might be wanting to do, it would be useful to know exactly. Thanks!

**Reply by Abhishek-1857** [reply]
const { Connection, LAMPORTS_PER_SOL, Keypair, PublicKey } = require('@solana/web3.js');
const { Contract, Program } = require('@solana/solidity');
const { readFileSync } = require('fs');

const FLIPPER_ABI = JSON.parse(readFileSync('./build/flipper.abi', 'utf8'));
const PROGRAM_SO = readFileSync('./build/bundle.so');

(async function () {
    console.log('Connecting to your local Solana node ...');
    const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
const payer=Keypair.generate()
// const user_address="7WHHG1yWXLaPgEphkm9TK7AQwjfScsAAotyKSyhhGnT2"
// const publicKey_user=new PublicKey(user_address)
console.log(payer.publicKey.toBase58())



    console.log('Airdropping SOL to a new wallet ...');
    const signature = await connection.requestAirdrop(payer.publicKey, LAMPORTS_PER_SOL);
    await connection.confirmTransaction(signature, 'confirmed');
    const balance=await connection.getBalance(payer.publicKey)
    console.log(balance)
    console.log(signature)

    const program = Keypair.generate();
    const storage = Keypair.generate();

    const contract = new Contract(connection, program.publicKey, storage.publicKey, FLIPPER_ABI, payer);

    await contract.load(program, PROGRAM_SO);

    console.log('Program deployment finished, deploying the flipper contract ...');

    await contract.deploy('flipper', [true], program, storage, 17);

    const res = await contract.functions.get();
    console.log('state: ' + res.result);

    await contract.functions.flip();

    const res2 = await contract.functions.get();
    console.log('state: ' + res2.result);
})();



This is the code from github repo of solang 
what i wanted is :
1. Use my devnet or testnet account which i have rather than using payer and generating random keys and accounts everytime i run my code.
2. I can get public key like this  
const user_address="7WHHG1yWXLaPgEphkm9TK7AQwjfScsAAotyKSyhhGnT2"
const publicKey_user=new PublicKey(user_address)
but what about the secret key how can i get that.
3. Is there any way i can get the address of contract deployed(  In this case i want my flipper contract to be deployed by my devnet account and get the address of deployed contract(if possible).

I am working on bridge and i need the contract address for the same.

**Reply by Abhishek-1857** [reply]
Thanks in advance for all your help...

**Reply by springrainroni** [reply]
when i deploy smart contract via solang on devnet it gives me contract.deploy function provides me that data but when i search 2C68dCxkXMuSsCbtYVFQ78QuNdQX9NAsvUz8EdeaUMWw this address on solana explorer but its not a program id. can you explain it better?
id {
  logs: [
    'Program 11111111111111111111111111111111 invoke [1]',
    'Program 11111111111111111111111111111111 success',
    'Program 2C68dCxkXMuSsCbtYVFQ78QuNdQX9NAsvUz8EdeaUMWw invoke [1]',
    'Program 2C68dCxkXMuSsCbtYVFQ78QuNdQX9NAsvUz8EdeaUMWw consumed 852 of 1400000 compute units',
    'Program 2C68dCxkXMuSsCbtYVFQ78QuNdQX9NAsvUz8EdeaUMWw success'
  ],
  events: [],
  computeUnitsUsed: 852
}
s

**Reply by seanyoung** [reply]
@Abhishek-1857 

1. Use my devnet or testnet account which i have rather than using payer and generating random keys and accounts everytime i run my code.

So rather than:

```
const payer = Keypair.generate();
```
You can write:
```
const payerKey = Uint8Array.from([
  202, 171, 192, 129, 150, 189, 204, 241, 142,  71, 205,
  2,  81,  97,   2, 176,  48,  81,  45,   1,  96, 138,
  220, 132, 231, 131, 120,  77,  66,  40,  97, 172,  91,
  245,  84, 221, 157, 190,   9, 145, 176, 130,  25,  43,
  72, 107, 190, 229,  75,  88, 191, 136,   7, 167, 109,
  91, 170, 164, 186,  15, 142,  36,  12,  23
]);

const payer = Keypair.fromSecretKey(secretKey);
```
The contents of the secret key can be what's in `~/.config/solana/id.json` or the output of `solana-keygen new --no-bip39-passphrase -o -`.

2. The public key is derived from the secret key. The secret/private key cannot be constructed from the public key (else encryption is broken)

3. The address of the program can be printed like so:

```
console.log(`program: ${program.publicKey.toBase58()}`);
```
Does that help?

**Reply by Abhishek-1857** [reply]
Yes thank you very much.... : ))) i was able to make a bridge  between eth to solana .....

**Reply by Abhishek-1857** [reply]
> when i deploy smart contract via solang on devnet it gives me contract.deploy function provides me that data but when i search 2C68dCxkXMuSsCbtYVFQ78QuNdQX9NAsvUz8EdeaUMWw this address on solana explorer but its not a program id. can you explain it better? id { logs: [ 'Program 11111111111111111111111111111111 invoke [1]', 'Program 11111111111111111111111111111111 success', 'Program 2C68dCxkXMuSsCbtYVFQ78QuNdQX9NAsvUz8EdeaUMWw invoke [1]', 'Program 2C68dCxkXMuSsCbtYVFQ78QuNdQX9NAsvUz8EdeaUMWw consumed 852 of 1400000 compute units', 'Program 2C68dCxkXMuSsCbtYVFQ78QuNdQX9NAsvUz8EdeaUMWw success' ], events: [], computeUnitsUsed: 852 } s

share your code so we can understand what you are doing.
